certora/specs/Vault.spec

```diff
@@ -0,0 +1,990 @@
+
+    uint256 balanceOfTreasury_ = underlying.balanceOf(treasury());
+    uint256 totalUnderlying_ = totalUnderlying();
+    uint256 accumulatedPerformanceFee_ = accumulatedPerfFee();
+
+    assert balanceOfTreasury_ - _balanceOfTreasury == _accumulatedPerformanceFee;
+    assert _totalUnderlying - totalUnderlying_ == _accumulatedPerformanceFee;
+    assert accumulatedPerformanceFee_ == 0;
+}
+
+
+
+/*
+    @Rule
+
+    @Category: state transition
+
+    @Description:
+        pause and exitPause functions should update states correctly and consistently
+*/
+rule integrity_of_pause_exitPause() {
+    require !paused();
+    env e;
+    pause(e);
+    assert paused();
+
+    env e1;
+    unpause(e1);
+    assert !paused();
+
+    require !exitPaused();
+    env e2;
+    exitPause(e2);
+    assert exitPaused();
+
+    env e3;
+    exitUnpause(e3);
+    assert !exitPaused();
+}
+
+
+/*
+    @Rule
+
+    @Category: unit test
+
+    @Description:
+        yield related calculations are correct
+*/
+rule yield_calculations_correct(address user) {
+    uint256 claimableYield = claimableYield(user);
+    uint256 claimableShares = claimableShares(user);
+    uint256 perfFee = perfFee(user);
+    uint256 yield = claimableYield + perfFee;
+    assert perfFee == pctOf(yield, perfFeePct());
+    assert yield > 0 => claimableShares > 0;
+    assert claimableShares == 0 => yield == 0;
+}
+
+
+/*
+    @Rule
+
+    @Category: unit test
+
+    @Description:
+        maxInvestableAmount equals investPct of totalUnderlying()
+*/
+rule maxInvestableAmount_correct() {
+    assert pctOf(totalUnderlying(), investPct()) == maxInvestableAmount();
+}
+
+
+/*
+    @Rule
+
+    @Category: unit test
+
+    @Description:
+        when the vault is in a loss, withdraw should revert
+*/
+rule withdraw_reverts_when_in_loss() {
+    require totalUnderlyingMinusSponsored() < totalPrincipal();
+    env e;
+    address to;
+    uint256[] ids;
+    withdraw@withrevert(e, to, ids); // should always revert
+    assert lastReverted;
+}
+
+
+/*
+    @Rule
+
+    @Category: unit test
+
+    @Description:
+        updateInvested function should make maxInvestableAmount equal alreadyInvested
+*/
+rule interity_of_updateInvested() {
+    env e;
+    require strategy() != currentContract;
+
+    underlying.setFee(e, 0);
+    uint256 _maxInvestableAmount = maxInvestableAmount();
+    uint256 _alreadyInvested = alreadyInvested();
+    require 
+        _maxInvestableAmount - _alreadyInvested > rebalanceMinimum() 
+        || 
+        _alreadyInvested - _maxInvestableAmount > rebalanceMinimum();
+
+    updateInvested(e);
+
+    uint256 maxInvestableAmount_ = maxInvestableAmount();
+    uint256 alreadyInvested_ = alreadyInvested();
+    assert maxInvestableAmount_ == alreadyInvested_;
+}
+
+
+/*
+    @Rule
+
+    @Category: unit test
+
+    @Description:
+        if a vault is paused, then deposit and sponsor functions will revert
+*/
+rule paused_vault_rejects_any_deposits() {
```

Comment:
This rule is not saying deposit functions ***only*** revert in this condition, it can revert in other conditions as other rules show. but this condition (paused) is sufficient to cause them to revert.